File: experiments/09-languages/src/go/bench.go
// E1, version Go (bibliothèque standard seule). Même plan de mesure, mêmes données et
// mêmes boucles « contrôlées » que les versions C, Rust et Python (README, section 4).
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"math"
	"os"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"
	"unsafe"
)

const (
	K         uint64 = 0x9e3779b97f4a7c15
	tableBits        = 17
	tableSize        = 1 << tableBits
	sysGetcpu        = 309 // getcpu sur linux/amd64, absent du paquet syscall
)

// ------------------------------------------------------------------ charges contrôlées

//go:noinline
func hashChain(u []uint64, h uint64) uint64 {
	for i := 0; i < len(u); i++ {
		h = (h ^ u[i]) * K
	}
	return h
}

//go:noinline
func dot(a, b []float64) float64 {
	s := 0.0
	for i := 0; i < len(a); i++ {
		s += a[i] * b[i]
	}
	return s
}

//go:noinline
func countTable(keys []uint32, slotKeys, slotCounts []uint32) {
	for i := 0; i < len(keys); i++ {
		stored := keys[i] + 1
		slot := uint32((uint64(stored) * K) >> (64 - tableBits))
		for slotKeys[slot] != 0 && slotKeys[slot] != stored {
			slot = (slot + 1) & (tableSize - 1)
		}
		slotKeys[slot] = stored
		slotCounts[slot]++
	}
}

// ------------------------------------------------------------------ versions idiomatiques

//go:noinline
func hashChainRange(u []uint64, h uint64) uint64 {
	for _, x := range u {
		h = (h ^ x) * K
	}
	return h
}

//go:noinline
func dotRange(a, b []float64) float64 {
	s := 0.0
	for i, x := range a {
		s += x * b[i]
	}
	return s
}

//go:noinline
func countMap(keys []uint32) map[uint32]uint32 {
	counts := make(map[uint32]uint32)
	for _, k := range keys {
		counts[k]++
	}
	return counts
}

// ------------------------------------------------------------------ outils

func cpu() int {
	var c uint32
	syscall.RawSyscall(sysGetcpu, uintptr(unsafe.Pointer(&c)), 0, 0)
	return int(c)
}

func xorshift(state *uint64) uint64 {
	*state ^= *state >> 12
	*state ^= *state << 25
	*state ^= *state >> 27
	return *state * 2685821657736338717
}

func peakRSSKiB() int64 {
	data, err := os.ReadFile("/proc/self/status")
	if err != nil {
		return -1
	}
	for _, line := range strings.Split(string(data), "\n") {
		if strings.HasPrefix(line, "VmHWM:") {
			v, _ := strconv.ParseInt(strings.Fields(line)[1], 10, 64)
			return v
		}
	}
	return -1
}

func numGC() uint32 {
	var stats runtime.MemStats
	runtime.ReadMemStats(&stats)
	return stats.NumGC
}

type measure struct {
	workload, variant string
	kind              int
	passes            int
}

const (
	hashControl = iota
	hashRange
	dotControl
	dotRange_
	countControl
	countMap_
)

func main() {
	if len(os.Args) == 2 && os.Args[1] == "--noop" {
		return
	}
	opt := map[string]string{}
	for i := 1; i+1 < len(os.Args); i += 2 {
		opt[os.Args[i]] = os.Args[i+1]
	}
	num := func(name string, def uint64) uint64 {
		if v, ok := opt[name]; ok {
			x, err := strconv.ParseUint(v, 10, 64)
			if err != nil {
				panic(err)
			}
			return x
		}
		return def
	}
	input, output, metaPath, build := opt["--input"], opt["--output"], opt["--meta"], opt["--build"]
	runID, seed := num("--run-id", 0), num("--seed", 20260922)
	rounds, warmup, scale := int(num("--rounds", 11)), int(num("--warmup-rounds", 1)), num("--scale-log2", 0)
	if input == "" || output == "" || metaPath == "" || runID == 0 || warmup >= rounds || scale > 12 {
		fmt.Fprintln(os.Stderr, "usage : bench --noop | --input F --output CSV --meta CSV --run-id N --build NOM")
		os.Exit(1)
	}

	nHash := 1 << (20 - scale)
	nDot, nCount := nHash, 1<<(22-scale)
	nTotal := max(nCount, 2*nDot)
	raw, err := os.ReadFile(input)
	if err != nil {
		panic(err)
	}
	u := make([]uint64, nTotal)
	for i := range u {
		u[i] = binary.LittleEndian.Uint64(raw[8*i:])
	}
	a, b := make([]float64, nDot), make([]float64, nDot)
	for i := 0; i < nDot; i++ {
		a[i] = float64(u[i]>>11) * 0x1p-53
		b[i] = float64(u[nDot+i]>>11) * 0x1p-53
	}
	keys := make([]uint32, nCount)
	var keysSum uint64
	for i := range keys {
		keys[i] = uint32(((u[i] >> 48) * ((u[i] >> 32) & 0xFFFF)) >> 16)
		keysSum += uint64(keys[i])
	}
	hashData := u[:nHash]
	slotKeys, slotCounts := make([]uint32, tableSize), make([]uint32, tableSize)

	measures := []measure{
		{"hash_chain", "control", hashControl, 1},
		{"hash_chain", "control_bis", hashControl, 1},
		{"hash_chain", "control_double", hashControl, 2},
		{"hash_chain", "idiomatic", hashRange, 1},
		{"dot", "control", dotControl, 1},
		{"dot", "idiomatic", dotRange_, 1},
		{"count", "control", countControl, 1},
		{"count", "idiomatic", countMap_, 1},
		{"count", "idiomatic_bis", countMap_, 1},
	}
	m := len(measures)
	state := seed ^ (runID * K)
	if state == 0 {
		state = 1
	}
	order := make([]int, 0, m*rounds)
	for r := 0; r < rounds; r++ {
		slice := make([]int, m)
		for j := range slice {
			slice[j] = j
		}
		for j := m; j > 1; j-- {
			other := int(xorshift(&state) % uint64(j))
			slice[j-1], slice[other] = slice[other], slice[j-1]
		}
		order = append(order, slice...)
	}

	type row struct {
		index, before, after int
		elapsed, result      uint64
		gc                   uint32
	}
	rows := make([]row, 0, len(order))
	firstCPU := cpu()
	gcStart := numGC()
	for _, index := range order {
		me := measures[index]
		if me.kind == countControl {
			clear(slotKeys)
			clear(slotCounts)
		}
		runtime.GC() // déchets des échantillons précédents collectés hors chronométrage
		gcBefore := numGC()
		before := cpu()
		t0 := time.Now()
		var r uint64
		var counts map[uint32]uint32
		switch me.kind {
		case hashControl:
			r = hashChain(hashData, 0)
			if me.passes == 2 {
				r = hashChain(hashData, r)
			}
		case hashRange:
			r = hashChainRange(hashData, 0)
		case dotControl:
			r = math.Float64bits(dot(a, b))
		case dotRange_:
			r = math.Float64bits(dotRange(a, b))
		case countControl:
			countTable(keys, slotKeys, slotCounts)
		case countMap_:
			counts = countMap(keys)
		}
		elapsed := uint64(time.Since(t0).Nanoseconds())
		after := cpu()
		gcAfter := numGC()
		if me.kind == countControl {
			for s, k := range slotKeys {
				if k != 0 {
					r += uint64(slotCounts[s]) * (uint64(k) * K)
				}
			}
		}
		if counts != nil {
			for k, c := range counts {
				r += uint64(c) * (uint64(k+1) * K)
			}
		}
		rows = append(rows, row{index, before, after, elapsed, r, gcAfter - gcBefore})
	}
	gcTotal := int(numGC()-gcStart) - len(order) // sans les collectes forcées

	f, err := os.Create(output)
	if err != nil {
		panic(err)
	}
	w := bufio.NewWriter(f)
	fmt.Fprintln(w, "run_id,build,seed,sample,round,phase,position,workload,variant,elements,passes,elapsed_ns,result,gc_delta,cpu_before,cpu_after")
	for s, rw := range rows {
		me := measures[rw.index]
		elements := nCount
		if me.kind == hashControl || me.kind == hashRange {
			elements = nHash
		} else if me.kind == dotControl || me.kind == dotRange_ {
			elements = nDot
		}
		phase := "measure"
		if s/m < warmup {
			phase = "warmup"
		}
		fmt.Fprintf(w, "%d,%s,%d,%d,%d,%s,%d,%s,%s,%d,%d,%d,%016x,%d,%d,%d\n", runID, build, seed, s, s/m, phase, s%m,
			me.workload, me.variant, elements, me.passes, rw.elapsed, rw.result, rw.gc, rw.before, rw.after)
	}
	if w.Flush() != nil || f.Close() != nil {
		os.Exit(1)
	}
	meta, err := os.Create(metaPath)
	if err != nil {
		panic(err)
	}
	fmt.Fprintln(meta, "run_id,build,runtime,first_cpu,peak_rss_kib,keys_sum,gc_total,jit,gil")
	fmt.Fprintf(meta, "%d,%s,%s,%d,%d,%d,%d,none,none\n", runID, build, runtime.Version(), firstCPU, peakRSSKiB(), keysSum, gcTotal)
	if meta.Close() != nil {
		os.Exit(1)
	}
}

File: experiments/09-languages/src/c/bench.c
#define _GNU_SOURCE

/* E1, version C. Même plan de mesure, mêmes données et mêmes boucles « contrôlées »
 * que les versions Rust, Go et Python (voir README, section 4). */

#include <errno.h>
#include <inttypes.h>
#include <sched.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define NOINLINE __attribute__((noinline))
#define K UINT64_C(0x9e3779b97f4a7c15)
#define TABLE_BITS 17
#define TABLE_SIZE (1u << TABLE_BITS)

/* ------------------------------------------------------------------ charges contrôlées */

static NOINLINE uint64_t hash_chain(const uint64_t *u, size_t n, uint64_t h) {
    for (size_t i = 0; i < n; ++i) h = (h ^ u[i]) * K;
    return h;
}

static NOINLINE double dot(const double *a, const double *b, size_t n) {
    double s = 0.0;
    for (size_t i = 0; i < n; ++i) s += a[i] * b[i];
    return s;
}

/* Table à adressage ouvert : clé + 1 stockée (0 = vide), sondage linéaire. */
static NOINLINE void count_table(const uint32_t *keys, size_t n, uint32_t *slot_keys, uint32_t *slot_counts) {
    for (size_t i = 0; i < n; ++i) {
        uint32_t stored = keys[i] + 1;
        uint32_t slot = (uint32_t)(((uint64_t)stored * K) >> (64 - TABLE_BITS));
        while (slot_keys[slot] != 0 && slot_keys[slot] != stored) slot = (slot + 1) & (TABLE_SIZE - 1);
        slot_keys[slot] = stored;
        slot_counts[slot] += 1;
    }
}

static uint64_t table_checksum(const uint32_t *slot_keys, const uint32_t *slot_counts) {
    uint64_t sum = 0;
    for (size_t s = 0; s < TABLE_SIZE; ++s)
        if (slot_keys[s] != 0) sum += (uint64_t)slot_counts[s] * ((uint64_t)slot_keys[s] * K);
    return sum;
}

/* ------------------------------------------------------------------ outils */

static uint64_t monotonic_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return (uint64_t)t.tv_sec * UINT64_C(1000000000) + (uint64_t)t.tv_nsec;
}

static uint64_t splitmix64(uint64_t *state) {
    uint64_t z = (*state += K);
    z = (z ^ (z >> 30)) * UINT64_C(0xbf58476d1ce4e5b9);
    z = (z ^ (z >> 27)) * UINT64_C(0x94d049bb133111eb);
    return z ^ (z >> 31);
}

static uint64_t xorshift(uint64_t *state) {
    *state ^= *state >> 12;
    *state ^= *state << 25;
    *state ^= *state >> 27;
    return *state * UINT64_C(2685821657736338717);
}

static uint64_t parse_u64(const char *text) {
    char *end = NULL;
    errno = 0;
    unsigned long long value = strtoull(text, &end, 10);
    if (errno || end == text || *end) {
        fprintf(stderr, "entier invalide : %s\n", text);
        exit(EXIT_FAILURE);
    }
    return value;
}

static long peak_rss_kib(void) {
    char line[256];
    long value = -1;
    FILE *f = fopen("/proc/self/status", "r");
    if (!f) return -1;
    while (fgets(line, sizeof line, f))
        if (sscanf(line, "VmHWM: %ld kB", &value) == 1) break;
    fclose(f);
    return value;
}

enum workload { W_HASH, W_DOT, W_COUNT };
struct measure { const char *workload, *variant; enum workload kind; int passes; };

int main(int argc, char **argv) {
    if (argc == 2 && strcmp(argv[1], "--noop") == 0) return 0;

    const char *input = NULL, *output = NULL, *meta_path = NULL, *generate = NULL, *build = "c";
    uint64_t run_id = 0, seed = 20260922, rounds = 11, warmup = 1, scale = 0, count = 0;
    for (int i = 1; i + 1 < argc; i += 2) {
        const char *o = argv[i], *v = argv[i + 1];
        if (!strcmp(o, "--input")) input = v;
        else if (!strcmp(o, "--output")) output = v;
        else if (!strcmp(o, "--meta")) meta_path = v;
        else if (!strcmp(o, "--generate")) generate = v;
        else if (!strcmp(o, "--build")) build = v;
        else if (!strcmp(o, "--run-id")) run_id = parse_u64(v);
        else if (!strcmp(o, "--seed")) seed = parse_u64(v);
        else if (!strcmp(o, "--rounds")) rounds = parse_u64(v);
        else if (!strcmp(o, "--warmup-rounds")) warmup = parse_u64(v);
        else if (!strcmp(o, "--scale-log2")) scale = parse_u64(v);
        else if (!strcmp(o, "--count")) count = parse_u64(v);
        else {
            fprintf(stderr, "option inconnue : %s\n", o);
            return EXIT_FAILURE;
        }
    }

    /* Données d'entrée communes : count entiers de 64 bits (SplitMix64), petit-boutistes. */
    if (generate) {
        FILE *f = fopen(generate, "wb");
        uint64_t state = seed;
        for (uint64_t i = 0; f && i < count; ++i) {
            uint64_t x = splitmix64(&state);
            if (fwrite(&x, sizeof x, 1, f) != 1) return EXIT_FAILURE;
        }
        return f && fclose(f) == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
    }
    if (!input || !output || !meta_path || run_id == 0 || warmup >= rounds || scale > 12) {
        fprintf(stderr, "usage : bench --noop | --generate FICHIER --count N --seed S | --input FICHIER "
                        "--output CSV --meta CSV --run-id N --build NOM [--seed S] [--rounds N] "
                        "[--warmup-rounds N] [--scale-log2 S]\n");
        return EXIT_FAILURE;
    }

    size_t n_hash = (size_t)1 << (20 - scale), n_dot = n_hash, n_count = (size_t)1 << (22 - scale);
    size_t n_total = n_count > 2 * n_dot ? n_count : 2 * n_dot;
    uint64_t *u = malloc(n_total * sizeof *u);
    double *a = malloc(n_dot * sizeof *a), *b = malloc(n_dot * sizeof *b);
    uint32_t *keys = malloc(n_count * sizeof *keys);
    uint32_t *slot_keys = calloc(TABLE_SIZE, sizeof *slot_keys), *slot_counts = calloc(TABLE_SIZE, sizeof *slot_counts);
    FILE *f = fopen(input, "rb");
    if (!u || !a || !b || !keys || !slot_keys || !slot_counts || !f || fread(u, sizeof *u, n_total, f) != n_total) {
        fprintf(stderr, "lecture de %s impossible\n", input);
        return EXIT_FAILURE;
    }
    fclose(f);

    /* Dérivations, identiques dans les quatre langages. */
    uint64_t keys_sum = 0;
    for (size_t i = 0; i < n_dot; ++i) {
        a[i] = (double)(u[i] >> 11) * 0x1.0p-53;
        b[i] = (double)(u[n_dot + i] >> 11) * 0x1.0p-53;
    }
    for (size_t i = 0; i < n_count; ++i) {
        keys[i] = (uint32_t)(((u[i] >> 48) * ((u[i] >> 32) & 0xFFFF)) >> 16);
        keys_sum += keys[i];
    }

    const struct measure measures[] = {
        {"hash_chain", "control", W_HASH, 1},
        {"hash_chain", "control_bis", W_HASH, 1},
        {"hash_chain", "control_double", W_HASH, 2},
        {"dot", "control", W_DOT, 1},
        {"count", "control", W_COUNT, 1},
    };
    size_t m = sizeof measures / sizeof measures[0];
    size_t total = m * rounds;
    size_t *order = malloc(total * sizeof *order);
    uint64_t *elapsed = malloc(total * sizeof *elapsed), *result = malloc(total * sizeof *result);
    int *cpu_before = malloc(total * sizeof *cpu_before), *cpu_after = malloc(total * sizeof *cpu_after);
    uint64_t state = seed ^ (run_id * K);
    if (state == 0) state = 1;
    for (size_t r = 0; r < rounds; ++r) {
        size_t *slice = order + r * m;
        for (size_t j = 0; j < m; ++j) slice[j] = j;
        for (size_t j = m; j > 1; --j) {
            size_t other = (size_t)(xorshift(&state) % j), t = slice[j - 1];
            slice[j - 1] = slice[other];
            slice[other] = t;
        }
    }

    int first_cpu = sched_getcpu();
    for (size_t s = 0; s < total; ++s) {
        const struct measure *me = &measures[order[s]];
        if (me->kind == W_COUNT) {   /* table remise à zéro hors chronométrage */
            memset(slot_keys, 0, TABLE_SIZE * sizeof *slot_keys);
            memset(slot_counts, 0, TABLE_SIZE * sizeof *slot_counts);
        }
        cpu_before[s] = sched_getcpu();
        uint64_t t0 = monotonic_ns(), r = 0;
        switch (me->kind) {
        case W_HASH:
            r = hash_chain(u, n_hash, 0);
            if (me->passes == 2) r = hash_chain(u, n_hash, r);
            break;
        case W_DOT: {
            double d = dot(a, b, n_dot);
            memcpy(&r, &d, sizeof r);
            break;
        }
        case W_COUNT:
            count_table(keys, n_count, slot_keys, slot_counts);
            break;
        }
        uint64_t t1 = monotonic_ns();
        cpu_after[s] = sched_getcpu();
        if (me->kind == W_COUNT) r = table_checksum(slot_keys, slot_counts);
        elapsed[s] = t1 - t0;
        result[s] = r;
    }

    FILE *out = fopen(output, "w");
    if (!out) return EXIT_FAILURE;
    fprintf(out, "run_id,build,seed,sample,round,phase,position,workload,variant,elements,passes,elapsed_ns,result,"
                 "gc_delta,cpu_before,cpu_after\n");
    for (size_t s = 0; s < total; ++s) {
        const struct measure *me = &measures[order[s]];
        size_t elements = me->kind == W_HASH ? n_hash : me->kind == W_DOT ? n_dot : n_count;
        fprintf(out, "%" PRIu64 ",%s,%" PRIu64 ",%zu,%zu,%s,%zu,%s,%s,%zu,%d,%" PRIu64 ",%016" PRIx64 ",0,%d,%d\n",
                run_id, build, seed, s, s / m, s / m < warmup ? "warmup" : "measure", s % m, me->workload,
                me->variant, elements, me->passes, elapsed[s], result[s], cpu_before[s], cpu_after[s]);
    }
    FILE *meta = fopen(meta_path, "w");
    if (!meta || fclose(out) != 0) return EXIT_FAILURE;
    fprintf(meta, "run_id,build,runtime,first_cpu,peak_rss_kib,keys_sum,gc_total,jit,gil\n");
    fprintf(meta, "%" PRIu64 ",%s,%s,%d,%ld,%" PRIu64 ",0,none,none\n", run_id, build, __VERSION__, first_cpu,
            peak_rss_kib(), keys_sum);
    return fclose(meta) == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
}
